С error when comparing aliased fixed arrays

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.3254987
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr/issues
vexe: /home/ge/.vlang/v
vexe mtime: 2024-12-05 14:09:51

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.1
Git vroot status: weekly.2024.48-67-g3254987b
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg comparison_of_aliased_fixed_arrays.vv`

```v
type Arr = [4]u8

fn (a Arr) u8_array_fixed() [4]u8 {
	return a
}

fn main() {
	a := Arr{} // Arr([0, 0, 0, 0])
	b := Arr{} // Arr([0, 0, 0, 0])
	println(a.u8_array_fixed()) // [0, 0, 0, 0]
	println(b.u8_array_fixed()) // [0, 0, 0, 0]
	println(a == b) // true

	// this workaround works
	a_ := a.u8_array_fixed()
	b_ := b.u8_array_fixed()
	println(a_ == b_) // true

	println(a.u8_array_fixed() == b.u8_array_fixed()) // C error here, expected: true
}
```

**What did you expect to see?**

'true' printed, no C errors.

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/comparison_of_aliased_fixed_arrays.01JEC061W2X53Q9PTR1PXSX9SK.tmp.c:5706: error: cannot convert 'unsigned char [4]' to 'struct _v_Array_fixed_u8_4'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzUxZWQ1MWQyZWY0Y2JjYzQ2ZjI5Y2QiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.pARb3igpf4u0HKiRzf9Ea8BNALEevkTG9PMT_zKip7o">Huly&reg;: <b>V_0.6-21511</b></a></sub>

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -2302,7 +2302,8 @@ pub:
 	typ    Type   // the type of the original expression
 	is_ptr bool   // whether the type is a pointer
 pub mut:
-	orig Expr // the original expression, which produced the C temp variable; used by x.str()
+	orig         Expr // the original expression, which produced the C temp variable; used by x.str()
+	is_fixed_ret bool // it is an array fixed returned from call
 }
 
 pub fn (node Node) pos() token.Pos {

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -426,13 +426,17 @@ fn (mut g Gen) gen_fixed_array_equality_fn(left_type ast.Type) string {
 	elem_info := left_typ.sym.array_fixed_info()
 	elem := g.unwrap(elem_info.elem_type)
 	size := elem_info.size
-	g.definitions.writeln('static bool ${ptr_styp}_arr_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
+	mut arg_styp := ptr_styp
+	if elem_info.is_fn_ret {
+		arg_styp = ptr_styp[3..] // removes the _v_ prefix for returning fixed array
+	}
+	g.definitions.writeln('static bool ${ptr_styp}_arr_eq(${arg_styp} a, ${arg_styp} b); // auto')
 
 	left := if left_type.has_flag(.option) { 'a.data' } else { 'a' }
 	right := if left_type.has_flag(.option) { 'b.data' } else { 'b' }
 
 	mut fn_builder := strings.new_builder(512)
-	fn_builder.writeln('static inline bool ${ptr_styp}_arr_eq(${ptr_styp} a, ${ptr_styp} b) {')
+	fn_builder.writeln('static inline bool ${ptr_styp}_arr_eq(${arg_styp} a, ${arg_styp} b) {')
 	fn_builder.writeln('\tfor (int i = 0; i < ${size}; ++i) {')
 	// compare every pair of elements of the two fixed arrays
 	if elem.sym.kind == .string {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3598,6 +3598,9 @@ fn (mut g Gen) expr(node_ ast.Expr) {
 		}
 		ast.CTempVar {
 			g.write(node.name)
+			if node.is_fixed_ret {
+				g.write('.ret_arr')
+			}
 		}
 		ast.DumpExpr {
 			g.dump_expr(node)

--- a/vlib/v/gen/c/ctempvars.v
+++ b/vlib/v/gen/c/ctempvars.v
@@ -12,18 +12,26 @@ fn (mut g Gen) new_ctemp_var(expr ast.Expr, expr_type ast.Type) ast.CTempVar {
 }
 
 fn (mut g Gen) new_ctemp_var_then_gen(expr ast.Expr, expr_type ast.Type) ast.CTempVar {
-	x := g.new_ctemp_var(expr, expr_type)
-	g.gen_ctemp_var(x)
+	mut x := g.new_ctemp_var(expr, expr_type)
+	g.gen_ctemp_var(mut x)
 	return x
 }
 
-fn (mut g Gen) gen_ctemp_var(tvar ast.CTempVar) {
+fn (mut g Gen) gen_ctemp_var(mut tvar ast.CTempVar) {
 	styp := g.styp(tvar.typ)
-	if g.table.final_sym(tvar.typ).kind == .array_fixed {
+	final_sym := g.table.final_sym(tvar.typ)
+	if final_sym.info is ast.ArrayFixed {
+		tvar.is_fixed_ret = final_sym.info.is_fn_ret
 		g.writeln('${styp} ${tvar.name};')
-		g.write('memcpy(&${tvar.name}, &')
-		g.expr(tvar.orig)
-		g.writeln(' , sizeof(${styp}));')
+		if tvar.is_fixed_ret {
+			g.write('memcpy(${tvar.name}.ret_arr, ')
+			g.expr(tvar.orig)
+			g.writeln(' , sizeof(${styp[3..]}));')
+		} else {
+			g.write('memcpy(&${tvar.name}, ')
+			g.expr(tvar.orig)
+			g.writeln(' , sizeof(${styp}));')
+		}
 	} else {
 		g.write('${styp} ${tvar.name} = ')
 		g.expr(tvar.orig)

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -1170,13 +1170,32 @@ fn (mut g Gen) gen_plain_infix_expr(node ast.InfixExpr) {
 			g.inside_interface_deref = inside_interface_deref_old
 		}
 	}
+	is_ctemp_fixed_ret := node.op in [.eq, .ne] && node.left is ast.CTempVar
+		&& node.left.is_fixed_ret
+	if is_ctemp_fixed_ret {
+		if node.op == .eq {
+			g.write('!')
+		}
+		g.write('memcmp(')
+	}
 	g.expr(node.left)
-	g.write(' ${node.op.str()} ')
+	if !is_ctemp_fixed_ret {
+		g.write(' ${node.op.str()} ')
+	} else {
+		g.write(', ')
+	}
+
+	if is_ctemp_fixed_ret {
+		g.write('(${g.styp(node.right_type)})')
+	}
 	if node.right_type.is_ptr() && node.right.is_auto_deref_var() && !node.left_type.is_pointer() {
 		g.write('*')
 		g.expr(node.right)
 	} else {
 		g.expr_with_cast(node.right, node.right_type, node.left_type)
+	}
+	if is_ctemp_fixed_ret {
+		g.write(', sizeof(${g.styp(node.right_type)}))')
 	}
 	if needs_cast {
 		g.write(')')

--- a/vlib/v/tests/aliases/alias_fixed_array_compare_test.v
+++ b/vlib/v/tests/aliases/alias_fixed_array_compare_test.v
@@ -1,0 +1,20 @@
+type Arr = [4]u8
+
+fn (a Arr) u8_array_fixed() [4]u8 {
+	return a
+}
+
+fn test_main() {
+	a := Arr{}
+	b := Arr{}
+	assert a.u8_array_fixed() == [u8(0), 0, 0, 0]!
+	assert b.u8_array_fixed() == [u8(0), 0, 0, 0]!
+	assert a == b // true
+
+	// // this workaround works
+	a_ := a.u8_array_fixed()
+	b_ := b.u8_array_fixed()
+	assert a_ == b_
+
+	assert a.u8_array_fixed() == b.u8_array_fixed()
+}

--- a/vlib/v/tests/aliases/alias_fixed_array_compare_test.v
+++ b/vlib/v/tests/aliases/alias_fixed_array_compare_test.v
@@ -9,9 +9,8 @@ fn test_main() {
 	b := Arr{}
 	assert a.u8_array_fixed() == [u8(0), 0, 0, 0]!
 	assert b.u8_array_fixed() == [u8(0), 0, 0, 0]!
-	assert a == b // true
+	assert a == b
 
-	// // this workaround works
 	a_ := a.u8_array_fixed()
 	b_ := b.u8_array_fixed()
 	assert a_ == b_

